src/inmanta/agent/executor.py

```diff
@@ -86,7 +129,21 @@ class ExecutorBlueprint(EnvBlueprint):
     sources: Sequence[ModuleSource]
     _hash_cache: Optional[str] = dataclasses.field(default=None, init=False, repr=False)
 
-    def generate_blueprint_hash(self) -> str:
+    @classmethod
+    def from_specs(cls, code: typing.Collection["ResourceInstallSpec"]) -> "ExecutorBlueprint":
+        sources = list({source for cd in code for source in cd.blueprint.sources})
+        requirements = list({req for cd in code for req in cd.blueprint.requirements})
+        pip_configs = [cd.blueprint.pip_config for cd in code]
+        base_pip = pip_configs[0]
+        for pip_config in pip_configs:
+            assert pip_config == base_pip, f"One agent is using multiple pipe configs: {base_pip} {pip_config}"
```

Comment:
it is a side effect of the structure of the api, but it can't happen, with the code as it currently is

Reply:
What I meant is, can't we do the assertion on the API boundary, and then lift the `pip_config` field out of `ResourceInstallSpec`. But now that I look at the structure again I see that that's infeasible, because `ResourceInstallSpec` is just a bunch of `ExecutorBlueprint` instances.